More work on a good development environment

danchr opened the pull request:
The general goal of this series is to make it people within Zimbra can easily use this appliance for building and working on Zimbra, including some of the less technical people.

The intent is that the changes should be in decreasing order of controversionalness (is that a word?) so please do cherry pick any changes that are OK.

I've tested this on Linux with LXC & libvirt as well as on OS X with Parallels and VirtualBox — the latter being _really_ slow. I haven't tried building any of the other modes; I'm not really sure how to test them?

danchr committed:
add support for Parallels & VMware Fusion on OS X

The VMware driver is proprietary, so I can't test it, but the
Parallels driver works. The only non-trivial change is removal of the
assumption that user directories reside in '/home'.

danchr committed:
add support for multicast dns

Originally introduced by Apple, Multicast DNS allows machines to
access each other as '$HOSTNAME.local' rather than numeric IP
addresess. It's pervasive on Macs, quite frequent on Linux, and
downloadable for Windows.

danchr committed:
allow customising memory allocation and CPU count

While at it, increase the default allocation to 4GB -- 2GB isn't quite
enough to build Zimbra.

danchr committed:
don't apply memory limit to LXC containers

Given that most of Zimbra is in Java, and Java is both extremely fond
of memory and extremely slow when it runs out, we shouldn't limit it
unless we really need to. When sharing a single kernel, individual
containers won't take up any more memory than they have to with a
higher limit, so the only effect of limiting memory is that it makes
everything slower.

At the very least, we should apply a higher limit; 2GB is not
sufficient for building Zimbra.

danchr committed:
add support for the vagrant-timezone plugin

UTC is just plain odd -- use the host timezone instead, when possible.

danchr committed:
default to development mode

The main motivation for this is targeting: The users least likely to
know about or understand Vagrant are most likely to want development
mode; the users wanting build mode generally know how to enable it.

danchr committed:
add startup scripts for ubuntu development instance

This includes a script that automatically starts any built and
installed instance of Zimbra, and one that adds /opt/zimbra/bin to $PATH.

danchr committed:
use pip rather than setuptools/easy_install

'pip' is the canonical way to install Python programs, these days.

danchr committed:
use local vsetup.sh script rather than one from GitHub

You'll need a full checkout anyway to build the appliance, so why not
use the 'vsetup.sh' right there on disk? Pointing elsewhere just has
the potential to confuse someone trying to modify the script.

plobbes commented:
LGTM!  I'll wait a little bit to see if @quanah has any concerns

quanah commented:
I would make it so git is always installed, regardless of type.

3rd party is nearly fully migrated to git, should be done no later than next week.
ZCS builds for FOSS are done entirely out of git via our git repo.  So essentially every use case requires git be available.

danchr committed:
README & config overhaul to describe the recent changes

danchr committed:
supply default boxes for known providers

This primary benefit is that you can build the appliance without
making any modifications to the source. You can still override the
defaults with VMBOX; the main benefit will be for users who don't want
to know about appliances.

danchr commented:
Okay, fixed a bug in the with the lxc provider, updated the config a bit, and updated the git patch as requested.

danchr committed:
install git on all boxes

Sooner or later, everyone will need it to fetch and/or build Zimbra.

danchr committed:
fix permissions for created directories

Turns out 'vagrant' isn't a member of the 'sudo' group in Ubuntu's
VirtualBox instance. Also, set the restricted deletion bit on any
created directories, so users don't accidentally delete anything
import in '/opt/zimbra'.
